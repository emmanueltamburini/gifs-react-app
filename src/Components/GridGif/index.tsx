import { useCallback, useEffect, useState } from "react";
import { getGif } from "../../helpers/getGif";
import { Gif } from '../../interfaces/Gif.interfaces';
import { GridItem } from "./GridItem/GridItem";

interface Props {
    category: string;
}

export const GridGif = ({ category }: Props) => {
    const [images, setImages] = useState<Gif[]>([]);

    const getImages = useCallback(
        async () => {
            const fetchImages = await getGif(category);
            setImages(fetchImages);
        },
        [category],
    )

    useEffect(() => {
        getImages();
    }, [getImages]);

    return (
        <>
            <h3>{category}</h3>
            <div className="card-grid">
                {images.map(image => {
                    return (
                        <GridItem key={image.id} image={image} />
                    );
                })}
            </div>
        </>
    );
}
