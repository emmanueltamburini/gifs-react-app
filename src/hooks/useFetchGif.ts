
import { useCallback, useEffect, useState } from "react";
import { getGif } from "../helpers/getGif";
import { Gif } from "../interfaces/Gif.interfaces";

export const useFetchGif = (category: string) => {
    const [images, setImages] = useState<Gif[]>([]);
    const [isLoading, setIsLoading] = useState(true);

    const getImages = useCallback(
        async () => {
            try {
                const fetchImages = await getGif(category);
                setImages(fetchImages);
                setIsLoading(false);
            } catch (error) {
                console.error(error);
                setIsLoading(false);
            }
        },
        [category],
    )

    useEffect(() => {
        getImages();
    }, [getImages]);

    return {images, isLoading};
}