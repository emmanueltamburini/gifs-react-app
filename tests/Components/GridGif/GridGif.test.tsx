import React from "react";
import { render, screen } from "@testing-library/react";
import { GridGif } from '../../../src/Components/GridGif/index';
import { useFetchGif } from "../../../src/hooks/useFetchGif";
import { Gif } from "../../../src/interfaces/Gif.interfaces";

jest.mock("../../../src/hooks/useFetchGif");

describe('Tests in GridGif', () => {
    const category = 'test';
    test('should be equal to snapshot', () => {
        const gifArray: Gif[] = [
            {
                id: '1',
                title: 'test-1',
                url: 'https://www.test.com/1'
            },
                        {
                id: '2',
                title: 'test-2',
                url: 'https://www.test.com/2'
            }
        ];

        (useFetchGif as jest.Mock).mockReturnValue({
            images: gifArray,
            isLoading: false,
        });

        const {container} = render(<GridGif category={category}></GridGif>);
        expect(container).toMatchSnapshot();
    });
    test('should show loading when init', () => {
        (useFetchGif as jest.Mock).mockReturnValue({
            images: [],
            isLoading: true,
        });

        render(<GridGif category={category}></GridGif>);
        expect(screen.getByText('Loading...'));
        expect(screen.getByText(category));
    });


    test('should show items when these are loaded from useFetchGif', () => {
        const gifArray: Gif[] = [
            {
                id: '1',
                title: 'test-1',
                url: 'https://www.test.com/1'
            },
                        {
                id: '2',
                title: 'test-2',
                url: 'https://www.test.com/2'
            }
        ];

        (useFetchGif as jest.Mock).mockReturnValue({
            images: gifArray,
            isLoading: false,
        });

        render(<GridGif category={category}></GridGif>);
        expect(screen.getAllByRole('img').length).toBe(2);
    })
});