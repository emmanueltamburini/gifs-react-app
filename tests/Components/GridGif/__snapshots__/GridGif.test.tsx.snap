// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in GridGif should be equal to snapshot 1`] = `
<div>
  <h3>
    test
  </h3>
  <div
    class="card-grid"
  >
    <div
      class="card"
    >
      <img
        alt="test-1"
        src="https://www.test.com/1"
      />
      <p>
        test-1
      </p>
    </div>
    <div
      class="card"
    >
      <img
        alt="test-2"
        src="https://www.test.com/2"
      />
      <p>
        test-2
      </p>
    </div>
  </div>
</div>
`;
